### PR TITLE
Appending in editor checks for duplicate images

### PR DESCRIPTION
**Pull request by apelsin21**
To fix issue  #7399 that appending maps in editor may duplicate images

## Checklist

- [x] Tested the change ingame
- [ ] Provided screenshots if it is a visual change
- [ ] Tested in combination with possibly related configuration options
- [ ] Written a unit test (especially base/) or added coverage to integration test
- [x] Considered possible null pointers and out of bounds array indexing
- [ ] Changed no physics that affect existing maps
- [ ] Tested the change with [ASan+UBSan or valgrind's memcheck](https://github.com/ddnet/ddnet/#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)

### CODE REVIEW

File: src/game/editor/editor.cpp [hunk 1]
```diff
@@ -8051,20 +8051,49 @@ bool CEditor::Append(const char *pFileName, int StorageType)
...
 	};
-	NewMap.ModifyImageIndex(s_ModifyAddIndex(m_Map.m_vpImages.size()));
+
 	NewMap.ModifySoundIndex(s_ModifyAddIndex(m_Map.m_vpSounds.size()));
 	NewMap.ModifyEnvelopeIndex(s_ModifyAddIndex(m_Map.m_vpEnvelopes.size()));
 
-	// transfer images
-	for(const auto &pImage : NewMap.m_vpImages)
-		m_Map.m_vpImages.push_back(pImage);
 	NewMap.m_vpImages.clear();
```
Comment (Robyt3) [inline src/game/editor/editor.cpp:8097]: Also move this line up, after the `for` loop, so all the image related code is all together in one block.

(and remove second empty line)

File: src/game/editor/editor.cpp [hunk 2]
```diff
@@ -8051,20 +8051,49 @@ bool CEditor::Append(const char *pFileName, int StorageType)
 	if(!NewMap.Load(pFileName, StorageType, std::move(ErrorHandler)))
 		return false;
 
+	static const auto &&s_replaceDuplicateImageIndex = [](int ToReplace, int ReplaceWith) {
```
Comment (Robyt3) [inline src/game/editor/editor.cpp:8054]: `s_ReplaceIndex`

File: src/game/editor/editor.cpp [hunk 3]
```diff
@@ -8051,20 +8051,49 @@ bool CEditor::Append(const char *pFileName, int StorageType)
...
+
+			dbg_msg("editor", "map contains image %s already, removing duplicate", pNewImage->m_aName);
+
+			//In the new map, replace the index of the duplicate image to the index of the same in the current map.
+			NewMap.ModifyImageIndex(s_replaceDuplicateImageIndex(IndexToReplace, IndexToReplaceWith));
+		}
+		else
+		{
+			NewMap.ModifyImageIndex(s_replaceDuplicateImageIndex(IndexToReplace, m_Map.m_vpImages.size()));
+			m_Map.m_vpImages.push_back(*NewMapIt);
```
Comment (Robyt3) [inline src/game/editor/editor.cpp:8082]: Use `pNewImage`

**Review by Robyt3 — CHANGES_REQUESTED** [review]
Anyone else got objections if we assume that images with the same name are duplicates or should we check for identical image data and rename the duplicate images if necessary? Same should apply to sounds.

**Comment by apelsin21** [comment]
Im thinking that this feature is only for convenience. Any maps that are submitted for official servers etc will be vetted in some way for images that are duplicates, so there's no reason to do a deeper comparison

Although ofc it would be nice to check sounds etc too

**Comment by heinrich5991** [comment]
> Anyone else got objections if we assume that images with the same name are duplicates or should we check for identical image data and rename the duplicate images if necessary?

Sounds ok to me.

**Comment by heinrich5991** [comment]
Rebased.

**Review by def- — APPROVED** [review]

**Pull request merged**